rts/motoko-rts/src/gc/mark_compact/mark_stack.rs

```diff
@@ -39,9 +39,9 @@ pub unsafe fn free_mark_stack() {
 }
 
 /// Doubles the stack size
-unsafe fn grow_stack() {
+unsafe fn grow_stack<M: Memory>(mem: &mut M) {
     let stack_cap: Words<u32> = STACK_BLOB_PTR.len().to_words();
-    let p = alloc_words(stack_cap).unskew() as *mut usize;
+    let p = mem.alloc_words(stack_cap).unskew() as *mut usize;
```

Comment:
A question about the existing code:
I see that `STACK_TOP` is expected to be equal to `p` in line 47, but how do we ensure that equality. Shouldn't we use `p` to set `STACK_TOP` or `STACK_PTR`?

Reply:
Here's how the Wasm heap looks like before starting compacting GC:

```
| static heap | dynamic heap |
^             ^              ^
0          heap base         hp
```

In compacting GC we first allocate a bitmap (size depends on dynamic heap size, does not grow) and a mark stack (size unknown, grows dynamically). At that point the heap looks like this:

```
| static heap | dynamic heap | bitmap | mark stack |
^             ^              ^        ^            ^
0          heap base         hp    stack base,  stack top
                                   stack ptr
```

As we push stuff to the mark stack stack ptr gets incremented towards stack top:

```
... | mark stack |
    ^    ^       ^
    |    p       |
stack base    stack top 
```

We grow the stack when `p == stack_top`, but we don't support moving the stack and don't have a linked-array representation, so the assumption here is that if we allocate more to grow the mark stack, new memory should be `stack top`. This effectively grows the mark stack without copying (moving) it. The assertion is checking that we didn't allocate anything after allocating the mark stack so `alloc_words` returned `stack top`.

What make sure that new allocation will be `stack top` is the rest of the GC code which does not allocate during marking. If we do any allocation during marking the assertion will fail.

@ulan Does this make sense?

Reply:
Thanks, got it! That's a nice trick. It will break if we switch to page-based layout. We would need non-contiguous stack at that point.